### PR TITLE
Second order gradient and Subgraph execution

### PR DESCRIPTION
**Pull request by piiswrong**

### CODE REVIEW

File: include/mxnet/ndarray.h [hunk 1]
```diff
@@ -875,8 +860,10 @@ class NDArray {
   size_t byte_offset_ = 0;
   /*! \brief type of data */
   int dtype_ = -1;
+  /*! \brief storage type of data */
+  NDArrayStorageType storage_type_ = kUndefinedStorage;
```
Comment (eric-haibin-lin) [inline include/mxnet/ndarray.h]: Is the `storage_type` in Chunk removed/no longer used?????

Comment (piiswrong) [reply include/mxnet/ndarray.h]: Still used. This is for accessing storage_type for none NDArray

**Comment by piiswrong** [comment]
@jermainewang @tqchen please have a look

File: tests/python/unittest/test_autograd.py [hunk 2]
```diff
@@ -117,7 +117,7 @@ def check_unary_func(x):
         f_square_grad = lambda x: [2*x]
         autograd_assert(x, func=f_square, grad_func=f_square_grad)
     uniform = nd.uniform(shape=(4, 5))
-    stypes = ['row_sparse', 'csr', 'default']
+    stypes = ['default']  #,'row_sparse', 'csr']
```
Comment (piiswrong) [inline tests/python/unittest/test_autograd.py]: @eric-haibin-lin Temporarily disabled sparse autograd. Please reenable after refactor

File: src/c_api/c_api_function.cc [hunk 3]
```diff
@@ -162,38 +162,35 @@ int MXCustomFunctionRecord(int num_inputs, NDArrayHandle *inputs,
                            MXCallbackList *callbacks) {
   using namespace mxnet;
   using namespace mxnet::custom_function;
-  using mxnet::autograd::AutogradRuntime;
```
Comment (eric-haibin-lin) [inline src/c_api/c_api_function.cc]: Is this function used only by custom ops registered in frontend?

File: tests/python/unittest/test_autograd.py (hunk 2 above)
Comment (eric-haibin-lin) [reply tests/python/unittest/test_autograd.py]: ok

File: src/operator/tensor/elemwise_binary_op_basic.cu [hunk 4]
```diff
@@ -36,21 +36,21 @@ NNVM_REGISTER_OP(_backward_add)
...
-NNVM_REGISTER_OP(_mul)
+NNVM_REGISTER_OP(elemwise_mul)
 .set_attr<FCompute>("FCompute<gpu>", ElemwiseBinaryOp::ComputeWithHalf2<gpu, mshadow::op::mul>);
 
 NNVM_REGISTER_OP(_backward_mul)
 .set_attr<FCompute>("FCompute<gpu>",
   ElemwiseBinaryOp::BackwardUseInWithHalf2<gpu, mshadow_op::right, mshadow_op::left>);
 
-NNVM_REGISTER_OP(_div)
+NNVM_REGISTER_OP(elemwise_div)
```
Comment (eric-haibin-lin) [inline src/operator/tensor/elemwise_binary_op_basic.cu]: Doesn't this break backward compatibility? @cjolivier01 @piiswrong

Comment (piiswrong) [reply src/operator/tensor/elemwise_binary_op_basic.cu]: This is fixing a bug introduced by chris. _div is renamed in .cc but not in .cu

File: src/c_api/c_api_function.cc (hunk 3 above)
Comment (piiswrong) [reply src/c_api/c_api_function.cc]: This is only used by autograd.Function

File: include/mxnet/imperative.h [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+   private:
+    struct CachedOpState {
+      std::vector<NDArray> buff;
+      std::vector<OpStatePtr> states;
+    };
+    std::mutex mutex_;
+    nnvm::Graph fwd_graph_;
+    nnvm::Graph grad_graph_;
+    nnvm::Graph full_graph_;
```
Comment (tqchen) [inline include/mxnet/imperative.h]: Why different graph object instead of one graph plus segment ?

Comment (piiswrong) [reply include/mxnet/imperative.h]: what segment? Depending on reqs, the gradient graph may need to be reconstructed

**Comment by jermainewang** [comment]
Is this the subgraph implementation? Or for second-order gradient?

File: include/mxnet/imperative.h (hunk 5 above)
Comment (tqchen) [reply include/mxnet/imperative.h]: i see, maybe add a comment here then why we need three graphs

File: src/c_api/c_api_ndarray.cc [hunk 6]
```diff
@@ -798,32 +341,58 @@ int MXAutogradBackward(mx_uint num_output,
...
 
   ograds.reserve(num_output);
   for (mx_uint i = 0; i < num_output; ++i) {
-    if (ograd_handles != nullptr && ograd_handles[i] != nullptr) {
-      ograds.emplace_back(*static_cast<NDArray*>(ograd_handles[i]));
+    if (ograd_handles != nullptr) {
+      ograds.emplace_back(reinterpret_cast<NDArray*>(ograd_handles[i]));
     } else {
-      ograds.emplace_back();
+      ograds.emplace_back(nullptr);
```
Comment (eric-haibin-lin) [inline src/c_api/c_api_ndarray.cc]: `ograds.emplace_back();` always generates dense ograd. Does `nullptr` respect the storage type inferred from the graph?

**Comment by piiswrong** [comment]
@jermainewang both

**Comment by jermainewang** [comment]
How will this integrate with the new column attribute design?

**Comment by sergeykolychev** [comment]
@piiswrong what is your estimation as of when this will be merged into master ?
I've finished on porting most of gluon interface to perl and of course this blocks me from filing the pull. I'm ok with whenever this goes to master, not meant to put a pressure on you, as I have  sparse ndarray to port now and that should take at least couple of weeks so I can wait and then combine both features in one pull.
Btw, when you have time can you please merge in https://github.com/apache/incubator-mxnet/pull/7815 ? It's gives perl interface nice performance boost (and passes perl tests).
Thanks!

**Comment by piiswrong** [comment]
@jermainewang Once the column store is merged you can refactor this by putting the attributes in the column store instead of plain graph attributes.

@sergeykolychev Should be in soonish

File: src/imperative/cached_op.cc [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+    g = nnvm::Graph();
+    g.outputs = fwd_graph_.outputs;
+    for (size_t i = 0; i < grad_graph_.outputs.size(); ++i) {
+      if (curr_grad_req_[i]) g.outputs.emplace_back(grad_graph_.outputs[i]);
+    }
+  }
+
+  const auto& idx = g.indexed_graph();
+  size_t num_forward_nodes = fwd_graph_.indexed_graph().num_nodes();
+  size_t num_forward_entries = fwd_graph_.indexed_graph().num_node_entries();
```
Comment (eric-haibin-lin) [inline src/imperative/cached_op.cc]: Is `num_forward_nodes` and `num_forward_entries ` from fwd_graph correct? The cached op takes the output of other ops and treats it as its input, which is not part of the fwd_graph inputs.

**Comment by kaylimekay** [comment]
Is there a list of operators that currently support the higher order gradients from grad()?

**Pull request merged**